core/embed/bootloader/messages.c

```diff
@@ -667,22 +707,49 @@ int process_msg_FirmwareUpload(uint8_t iface_num, uint32_t msg_size,
     return UPLOAD_ERR_INVALID_CHUNK_HASH;
   }
 
-  ensure(flash_unlock_write(), NULL);
-
-  const uint32_t *const src = (const uint32_t *const)CHUNK_BUFFER_PTR;
+  const uint32_t *quadword_ptr = (const uint32_t *)CHUNK_BUFFER_PTR;
+  uint32_t bytes_remaining = chunk_size;
+  uint32_t write_offset = firmware_block * IMAGE_CHUNK_SIZE;
 
-  for (int i = 0; i < chunk_size / (sizeof(uint32_t) * 4); i++) {
-    ensure(flash_area_write_quadword(
-               &FIRMWARE_AREA,
-               firmware_block * IMAGE_CHUNK_SIZE + i * 4 * sizeof(uint32_t),
-               &src[4 * i]),
-           NULL);
+  while (bytes_remaining > 0) {
+    uint32_t bytes_erased = 0;
+    if (write_offset >= erase_offset) {
+      ensure(
+          flash_area_erase_partial(&FIRMWARE_AREA, erase_offset, &bytes_erased),
+          NULL);
+      erase_offset += bytes_erased;
+    } else {
+      bytes_erased = erase_offset - write_offset;
+    }
+    uint32_t bytes_to_write = MIN(bytes_erased, bytes_remaining);
+    uint32_t write_end = write_offset + bytes_to_write;
+    ensure(flash_unlock_write(), NULL);
+    while (write_offset < write_end) {
+      // write quad word (16 bytes) to the flash
+      ensure(
+          flash_area_write_quadword(&FIRMWARE_AREA, write_offset, quadword_ptr),
+          NULL);
+      write_offset += 4 * sizeof(uint32_t);
+      quadword_ptr += 4;
+    }
+    ensure(flash_lock_write(), NULL);
+    bytes_remaining -= bytes_to_write;
   }
 
```

Comment:
this piece of code is rather opaque and needs significantly more comments

Reply:
I've added some comments, see  d18d6e46ac3bb225ba483ecec148a552216d441e